File: leetcode/6_recursion_and_looping/440_K-th_Smallest_in_Lexicographical_Order.py
"""
Given two integers n and k, return the kth lexicographically smallest integer in the range [1, n].



Example 1:

Input: n = 13, k = 2
Output: 10
Explanation: The lexicographical order is [1, 10, 11, 12, 13, 2, 3, 4, 5, 6, 7, 8, 9], so the second smallest number is 10.

Example 2:

Input: n = 1, k = 1
Output: 1



Constraints:

    1 <= k <= n <= 109
"""


class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        def _count(cur: int) -> int:
            res = 0
            nei = cur + 1
            while cur <= n:
                res += min(nei, n + 1) - cur
                cur *= 10
                nei *= 10
            return res

        cur = 1
        i = 1

        while i < k:
            steps = _count(cur)
            if i + steps <= k:
                cur = cur + 1
                i += steps
            else:
                cur *= 10
                i += 1

        return cur

File: leetcode/6_recursion_and_looping/440_K-th_Smallest_in_Lexicographical_Order.go
func findKthNumber(n int, k int) int {
	curr := 1
	k -= 1
	for k > 0 {
		count := count(curr, curr+1, n)
		if count <= k {
			curr += 1
			k -= count
		} else {
			curr *= 10
			k--
		}
	}

	return curr
}

func count(curr, next int, n int) int {
	count := 0
	for curr <= n {
		count += next - curr
		curr *= 10
		next *= 10

		next = min(next, n+1)
	}

	return count
}
